### PR TITLE
JaxModel (For integerating Jax into deepchem)

### PR DESCRIPTION
**Pull request by VIGNESHinZONE**
As part of my GSOC project, 

Here is my first PR on the JaxModel, Different methods of JaxModel need to interact with each other `Save and Load Models` and hence will be implemented slowly,

- [x] Finished the constructor method `init`
- [x] Finished most of the `fit` method
- [x] Accommodate the output types for both regression & classification. 
- [x] Documentation of current work and future development 
- [x] Adding Test suites for the `fit` method
- [x] Adding `jax` libraries in `requirements.txt`  
 
 ---
About the PR, 

JaxModel can be reproduced in this [google collab](https://colab.research.google.com/drive/1apfqWdYrIgohje_dnAd0WAUCAeBQTH08#scrollTo=Wi4BR34fgG2l)

Will add the `requirements.txt` once CI gets sorted.

Mentors: @rbharath @ncfrey

### CODE REVIEW

File: deepchem/models/jax_models/jax_model.py [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+import optax
+
+logger = logging.getLogger(__name__)
+
+
+class JaxModel(Model):
+  """This is a DeepChem model implemented by a Jax Model
+
+  Here is a simple example of that uses JaxModel to train a
+  Haiku (JAX Nueral Network Library) based model on deepchem
```
Comment (rbharath) [inline deepchem/models/jax_models/jax_model.py:29]: typo: Neural

File: deepchem/models/jax_models/jax_model.py [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+  >> x, _, _, _ = next(iter(train_dataset.iterbatches(batch_size=256)))
+  >> params = model.init(rng, x)
+  >> j_m = JaxModel(model, params, 256, 0.001, 100)
+  >> j_m.fit(train_dataset)
+
+  All optimizations will be done using the optax library.
+  """
+
+  def __init__(self,
+               model,
```
Comment (rbharath) [inline deepchem/models/jax_models/jax_model.py:46]: Type signature for `model` needed

File: deepchem/models/jax_models/jax_model.py [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+
+from deepchem.data import Dataset
+from deepchem.models.models import Model
+from deepchem.models.losses import Loss
+from deepchem.models.optimizers import Optimizer
+from typing import Any, Callable, Iterable, List, Optional, Tuple, Union
+from deepchem.utils.typing import LossFn
+
+# JAX depend
+import jax.numpy as jnp
```
Comment (rbharath) [inline deepchem/models/jax_models/jax_model.py:17]: These should be wrapped in try/catch guards and optional dependencies

**Review by rbharath — COMMENTED** [review]
Did a first pass through and design looks reasonable so far. Have a few housekeeping comments.

@peastman Would you mind taking a quick look if possible? It would be good to get your thoughts on the design so far as well

File: deepchem/models/jax_models/jax_model.py (hunk 2 above)
Comment (VIGNESHinZONE) [reply deepchem/models/jax_models/jax_model.py:46]: I still haven't decided on the model signature, because Jax in its true forms uses pure functions to calculate gradients. Hence mostly likely I will call it `Callable` type. 

There are many popular Jax Neural Network Libraries like [`Haiku`](https://github.com/deepmind/dm-haiku) and [`Flax`](https://github.com/google/flax), and im trying to build it library agnostic. In my current experiments, I have been using `Haiku` models.

**Comment by VIGNESHinZONE** [comment]
@rbharath @ncfrey 

1.  The CI does not have any installation faults in ubuntu/linux 3.7 and test functions are also passing. Only the 1 mypy error, which im not sure how to tackle because its under  `python3.7/site-packages`.
2. Added the `warning messages` we had talked about yesterday.

**Comment by VIGNESHinZONE** [comment]
Here is this one strange `Mypy` error caused in `python/site-packages/`, are there ways to bypass this error from our end?
```
/usr/share/miniconda3/envs/deepchem/lib/python3.7/site-packages/jaxlib/xla_extension-stubs/__init__.pyi:100: error: invalid syntax

```
cc: @peastman

File: deepchem/models/jax_models/jax_model.py [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+from deepchem.utils.typing import LossFn
+
+# JAX depend
+try:
+  import jax.numpy as jnp
+  import jax
+  import haiku as hk
+  import optax
+except:
+  raise ImportError("This class requires Jax, haiku and optax installed.")
```
Comment (ncfrey) [inline deepchem/models/jax_models/jax_model.py:23]: "to be installed"

File: deepchem/models/jax_models/jax_model.py [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+class JaxModel(Model):
+  """This is a DeepChem model implemented by a Jax Model
+
+  Here is a simple example of that uses JaxModel to train a
+  Haiku (JAX Neural Network Library) based model on deepchem
+  dataset.
```
Comment (ncfrey) [inline deepchem/models/jax_models/jax_model.py:35]: Can we add a bit more documentation here summarizing the 3 points you brought up during the meeting that discuss why a user would be interested in using a `JaxModel`?

The easy distributed training in particular seems like a great advantage to me.

File: deepchem/models/jax_models/jax_model.py (hunk 2 above)
Comment (ncfrey) [reply deepchem/models/jax_models/jax_model.py:51]: `Callable` sounds good here. Could also do `Union[...,...]` with each framework that is supported.

File: deepchem/models/jax_models/jax_model.py [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+                                Optimizer] = optax.adam(1e-3),
+               log_frequency: int = 100,
+               **kwargs):
+    """
+    Create a new JaxModel
+
+    Parameters
+    ----------
+    model: hk.State or Function
+      Any Jax based model that has a `apply` method for computing the network.
```
Comment (ncfrey) [inline deepchem/models/jax_models/jax_model.py:67]: Can mention that haiku, flax, etc. are supported here.

File: deepchem/models/jax_models/jax_model.py [hunk 7]
```diff
@@ -0,0 +1,373 @@
...
+
+    Parameters
+    ----------
+    dataset: Dataset
+      the Dataset to train on
+    nb_epoch: int
+      the number of epochs to train for
+    deterministic: bool
+      if True, the samples are processed in order.  If False, a different random
+      order is used for each epoch.
```
Comment (ncfrey) [inline deepchem/models/jax_models/jax_model.py:167]: Is this fixing the seed so that the batches are exactly the same between training runs?

File: deepchem/models/jax_models/jax_model.py [hunk 8]
```diff
@@ -0,0 +1,373 @@
...
+  >> params = model.init(rng, x)
+  >> j_m = JaxModel(model, params, 256, 0.001, 100)
+  >> j_m.fit(train_dataset)
+
+  All optimizations will be done using the optax library.
+  """
+
+  def __init__(self,
+               model,
+               params: hk.Params,
```
Comment (ncfrey) [inline deepchem/models/jax_models/jax_model.py:52]: Should this also be a `Union`?

File: deepchem/models/jax_models/jax_model.py [hunk 9]
```diff
@@ -0,0 +1,373 @@
...
+      updates, opt_state = optimizer.update(grads, opt_state)
+      new_params = optax.apply_updates(params, updates)
+      return new_params, opt_state, batch_loss
+
+    return update
+
+  def _prepare_batch(self, batch):
+    inputs, labels, weights = batch
+    inputs = [
+        x.astype(np.float32) if x.dtype == np.float64 else x for x in inputs
```
Comment (ncfrey) [inline deepchem/models/jax_models/jax_model.py:316]: Does mixing `jnp` and `np` matter at all?

File: deepchem/models/jax_models/jax_model.py [hunk 10]
```diff
@@ -0,0 +1,373 @@
...
+      whether to pad each batch up to this model's preferred batch size
+
+    Returns
+    -------
+    a generator that iterates batches, each represented as a tuple of lists:
+    ([inputs], [outputs], [weights])
+    """
+
+    for epoch in range(epochs):
+      for (X_b, y_b, w_b, ids_b) in dataset.iterbatches(
```
Comment (ncfrey) [inline deepchem/models/jax_models/jax_model.py:369]: Is `ids_b` used for anything?

File: deepchem/models/tests/test_jax_model.py [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+  import optax
+  from deepchem.models import JaxModel
+  has_haiku_and_optax = True
+except:
+  has_haiku_and_optax = False
+
+
+@unittest.skipIf(not has_haiku_and_optax,
+                 'Jax, Haiku, or Optax are not installed')
+def test_jax_model_for_regression():
```
Comment (ncfrey) [inline deepchem/models/tests/test_jax_model.py:19]: Should we add a `TODO` comment here on additional methods that will be tested after they're written?

File: env.common.yml [hunk 12]
```diff
@@ -29,3 +29,5 @@ dependencies:
     - transformers==4.6.*
     - xgboost==1.*
     - git+https://github.com/samoturk/mol2vec
+    - git+https://github.com/deepmind/dm-haiku
```
Comment (rbharath) [inline env.common.yml:32]: Can you add these dependencies to the soft dependencies list in the docs in `docs/source/get_started/requirements.rst`?

File: deepchem/models/jax_models/jax_model.py [hunk 13]
```diff
@@ -0,0 +1,373 @@
...
+      whether to log progress to Weights & Biases during training
+
+    Work in Progress
+    ----------------
+    [1] Integerate the optax losses, optimizers, schedulers with Deepchem
+    [2] Support for saving & loading the model.
+    """
+    super(JaxModel, self).__init__(model=model, **kwargs)
+    warnings.warn(
+        'JaxModel is still under work and hence a lot of the method might not be implemented'
```
Comment (rbharath) [inline deepchem/models/jax_models/jax_model.py:102]: How about "JaxModel is still in active development and all features may not yet be implemented"

**Review by rbharath — COMMENTED** [review]
Looks pretty good. Have a few comments about documentation below

File: env.common.yml (hunk 12 above)
Comment (peastman) [reply env.common.yml:32]: Is it really necessary to add these dependencies?  Our dependency matrix is already reaching a point where installing DeepChem can be impossible.  We've had to put off our next release because of it.  Every extra dependency makes the problem even worse.  We desperately need to be getting rid of dependencies, not adding new ones.

Comment (rbharath) [reply env.common.yml:32]: Perhaps we should add a `deepchem-jax` package as we've previously discussed instead of adding on to the dependencies of the base package. It's definitely true that the dependency matrix for deepchem has started to become unsustainable.

Comment (ncfrey) [reply env.common.yml:32]: +1 for a `deepchem-jax` package that doesn't include PyTorch, Tensorflow, dgl, transformers, etc. Would vastly simplify the dependencies here.

File: deepchem/models/jax_models/jax_model.py (hunk 7 above)
Comment (peastman) [reply deepchem/models/jax_models/jax_model.py:167]: This option isn't to get the same behavior between runs.  To do that, you would fix the random number seeds.  I'm honestly not sure when someone would want to set this option to True.  Using a different order for each epoch usually gives better training.  Maybe there's some situation I haven't thought of where someone would want the same order for every epoch.

Comment (VIGNESHinZONE) [reply deepchem/models/jax_models/jax_model.py:167]: 1) Oh my bad, thanks for letting me know. I had gotten confused with `seed` value being passed here https://github.com/deepchem/deepchem/blob/66402d1cf34d51c13ed491f1f14eaffca133dfab/deepchem/splits/splitters.py#L369

2) Yes, I agree that having different orders of batches during training will definitely give better training. The precise results of a model can only be reproduced if the random training order of batches and initial model weights are the same. With my limited exposure to `Jax`, it has an amazing ability to reproduce results because of the way in which it generates its random numbers. We were discussing this, merely to check this ability to reproduce very well in Jax. In my runs, I wasn't able to reproduce my results on different instances of collab sessions.

Based on your suggestions, I think setting `random.random(seed)` might fix the problem.

Comment (peastman) [reply deepchem/models/jax_models/jax_model.py:167]: And also `np.random.seed()`.  Actually, I don't think we ever use `random` anywhere.  But we do use `np.random`.

File: deepchem/models/jax_models/jax_model.py (hunk 5 above)
Comment (VIGNESHinZONE) [reply deepchem/models/jax_models/jax_model.py:35]: Yeah sure, I will add them. But i think there is a lot more to add in this section and hoping to add it in the end.

File: deepchem/models/jax_models/jax_model.py (hunk 9 above)
Comment (VIGNESHinZONE) [reply deepchem/models/jax_models/jax_model.py:316]: well `jnp` can perform all the operations that are performed in `np`, hence its not a necessary.

**Comment by VIGNESHinZONE** [comment]
@rbharath Jax tests are passing successfully, hence we could move ahead with this one. The only thing which needs to be noted is the Model inputs depending on library (Haiky, Flax, or pure Jax) will change and this design will change in the subsequent PRs.

File: deepchem/models/jax_models/jax_model.py [hunk 14]
```diff
@@ -0,0 +1,371 @@
...
+
+
+class JaxModel(Model):
+  """This is a DeepChem model implemented by a Jax Model
+
+  Here is a simple example of that uses JaxModel to train a
+  Haiku (JAX Neural Network Library) based model on deepchem
+  dataset.
+
+  >> def f(x):
```
Comment (rbharath) [inline deepchem/models/jax_models/jax_model.py:34]: Since you're using `>>` instead of `>>>` this test won't run during doctests. Is that on purpose?

File: deepchem/models/jax_models/jax_model.py [hunk 15]
```diff
@@ -0,0 +1,371 @@
...
+    model_dir: str, optional (default None)
+      Will be added along with the save & load method
+    tensorboard: bool, optional (default False)
+      whether to log progress to TensorBoard during training
+    wandb: bool, optional (default False)
+      whether to log progress to Weights & Biases during training
+
+    Work in Progress
+    ----------------
+    [1] Integerate the optax losses, optimizers, schedulers with Deepchem
```
Comment (rbharath) [inline deepchem/models/jax_models/jax_model.py:95]: Typo: integrate

File: requirements/jax/env_jax.yml [hunk 16]
```diff
@@ -4,5 +4,11 @@ channels:
   - conda-forge
   - defaults
 dependencies:
-  - numpy==1.19.*
+  - numpy==1.21.*
+  - rdkit
```
Comment (rbharath) [inline requirements/jax/env_jax.yml:8]: Should rdkit be a dependency here?

**Review by rbharath — APPROVED** [review]
I have a few minor comments and I think there are still a couple of older comments that haven't been fully responded to. That said, I think these are all minor and we can go ahead and merge this PR once you're ready @VIGNESHinZONE. Could you let me know once all tests for this PR are passing?

File: deepchem/models/jax_models/jax_model.py (hunk 14 above)
Comment (VIGNESHinZONE) [reply deepchem/models/jax_models/jax_model.py:34]: Yes it was intentional, doctest are run in `deepchem core` workflow where I haven't included `jax` dependency.

File: requirements/jax/env_jax.yml (hunk 16 above)
Comment (VIGNESHinZONE) [reply requirements/jax/env_jax.yml:8]: In my tests, I'm using `delaney` & `bace` datasets from `molnet` which uses circular fingerprints.

File: deepchem/models/jax_models/jax_model.py (hunk 14 above)
Comment (VIGNESHinZONE) [reply deepchem/models/jax_models/jax_model.py:34]: I think, now that we have mostly sorted out the CI, i think i can add `Jax` in here https://github.com/deepchem/deepchem/blob/master/env.common.yml

**Comment by VIGNESHinZONE** [comment]
@rbharath I have added few fixes and all the `Jax` tests are passing.

File: deepchem/models/jax_models/jax_model.py (hunk 14 above)
Comment (rbharath) [reply deepchem/models/jax_models/jax_model.py:34]: Sounds reasonable! We shouldn't add Jax to common though I think since that's the platform free core for DeepChem
